scripts/saml-configuration

```diff
@@ -0,0 +1,151 @@
+try:
+    idps: Dict[str, Any] = settings.SOCIAL_AUTH_SAML_ENABLED_IDPS
+except Exception:
+    logging.critical('Cannot receive Zulip SAML configuration')
+    traceback.print_exc()
+    sys.exit(1)
+
+for idp in idps:
+    # The list of the IdPs signing certificates.
+    certs_list: List[str] = []
+    new_cert: str
+    new_cert_object: x509.Certificate
+
+    # Check the provided settings.
+    if 'metadata_url' not in idps[idp] or 'entity_id' not in idps[idp]:
+        logging.error('No \'metadata_url\' and/or \'entity_id\' field for IdP '
+                      '\'%s\' in the SAML configuration', idp)
+        continue
+
+    # Get the data from the metadata url of the IdP.
+    data: Dict[str, Any] = Parser.parse_remote(
+        idps[idp]['metadata_url'],
+        validate_cert = True,
+        entity_id = idps[idp]['entity_id']
+    )
+
+    if 'idp' not in data:
+        logging.error('Cannot receive metadata for IdP ' + idp)
+        logging.debug(data)
+        continue
+
+    # Some documentation:
+    # - https://github.com/onelogin/python3-saml
+    # - https://github.com/onelogin/python3-saml/blob/master/src\
+    #     /onelogin/saml2/idp_metadata_parser.py
+
+    if 'x509cert' in data['idp']:
+        # A single certificate, used for both signing and encryption.
+        certs_list.append(data['idp']['x509cert'])
+    elif ('x509certMulti' in data['idp']
+          and 'signing' in data['idp']['x509certMulti']):
+        # There are multiple certificates, maybe different ones for
+        # signing and encryption.
+        certs_list.extend(data['idp']['x509certMulti']['signing'])
+
+    if not certs_list:
+        logging.error('Cannot find any certificates for IdP ' + idp)
+        continue
+
+    # Format the certificates properly.
+    # See https://github.com/onelogin/python3-saml/blob/master/src\
+    #   /onelogin/saml2/utils.py
+    certs: Iterator[str] = map(Utils.format_cert, certs_list)
+
+    # Map the certificates to their x509 objects in order to be able to
+    # determine the validity of the certificate.
+    # See https://cryptography.io/en/latest/x509/reference.html\
+    #   #cryptography.x509.Certificate
+    certs_objects: Iterator[Tuple[str, x509.Certificate]] = map(
+        lambda cert: (cert, x509.load_pem_x509_certificate(
+            cert.encode(), backend = default_backend()
+        )),
+        certs
+    )
+
+    # Get the first currently valid certificate.
+    try:
+        new_cert, new_cert_object = next(filter(
+            lambda cert_object: is_valid(cert_object[1]), certs_objects
+        ))
+    except StopIteration:
+        logging.error('Cannot find any valid certificate for IdP ' + idp)
+        logging.debug(traceback.format_exc())
+        continue
+
+    # Check for the best certificate available:
+    #  - It is currently valid.
+    #  - It is valid for as long as possible.
+    for cert, cert_object in certs_objects:
```

Comment:
Does this intentionally exclude the items of the `certs_objects` iterator that have already been consumed above?